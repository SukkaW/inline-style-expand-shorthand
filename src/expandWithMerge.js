import expandProperty from './expandProperty'

function mergeBase(longhands, base) {
  for (const property in longhands) {
    if (base.hasOwnProperty(property)) {
      longhands[property] = base[property]
    }
  }

  return longhands
}

const priority = [
  'borderLeft',
  'borderRight',
  'borderBottom',
  'borderTop',
  'borderWidth',
  'borderStyle',
  'borderColor',
]

export default function expandWithMerge(style) {
  // we reverse the key order to make sure more specific properties
  // will always overwrite the unspecific shorthands
  const sortedKeys = Object.keys(style)
    .sort((a, b) =>
      priority.indexOf(a) && priority.indexOf(b)
        ? priority.indexOf(a) > priority.indexOf(b)
        : a > b || -1
    )
    .reverse()

  for (const property of sortedKeys) {
    const value = style[property]

    if (typeof value === 'string' || typeof value === 'number') {
      const expansion = expandProperty(property, value)

      if (expansion) {
        Object.assign(style, mergeBase(expansion, style))
        delete style[property]
      }
    } else if (value === null) {
      // should skip
    } else if (Array.isArray(value)) {
      if (property === 'extend') {
        value.map(expandWithMerge)
      } else {
        const expansion = expandProperty(property, value)

        if (expansion) {
          Object.assign(style, mergeBase(expansion, style))
          delete style[property]
        }
      }
    } else if (typeof value === 'object') {
      expandWithMerge(value)
    }
  }

  return style
}
