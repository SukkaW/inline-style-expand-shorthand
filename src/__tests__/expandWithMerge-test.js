import expandWithMerge from '../expandWithMerge'

describe('Expanding (with merge) style objects', () => {
  it('should expand and merge properties in nested objects', () => {
    expect(
      expandWithMerge({
        padding: '20px 10px',
        paddingBottom: '15px',
        ':hover': {
          borderWidth: '5px 2px 3px 4px',
          borderBottomWidth: '10px',
        },
      })
    ).toMatchSnapshot()
  })

  it('should expand values in arrays', () => {
    expect(
      expandWithMerge({
        numeralArray: [1, 2],
        padding: [1, '20px'],
        extend: [
          { padding: '10px' },
          { margin: 0 },
          { outline: [0, '10px solid red'], outlineColor: 'green' },
        ],
      })
    ).toMatchSnapshot()
  })

  it('should expand and merge properties with non-undefined values', () => {
    expect(
      expandWithMerge({
        padding: '20px 10px',
        paddingBottom: 0,
        paddingLeft: null,
        paddingRight: undefined,
      })
    ).toMatchSnapshot()
  })

  it('should expand and merge border properties in nested objects', () => {
    expect(
      expandWithMerge({
        border: '1px solid grey',
        borderLeft: '2px dotted black',
      })
    ).toMatchSnapshot()
  })

  it('should expand and merge multiple border properties in nested objects', () => {
    expect(
      expandWithMerge({
        border: '1px solid grey',
        borderLeft: '2px dotted black',
        borderWidth: '5px 6px 7px 8px',
        borderLeftWidth: '10px',
      })
    ).toMatchSnapshot()
  })
})
