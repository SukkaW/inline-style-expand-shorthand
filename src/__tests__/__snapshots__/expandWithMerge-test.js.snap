// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Expanding (with merge) style objects should expand and merge border properties in nested objects 1`] = `
Object {
  "borderBottomColor": "grey",
  "borderBottomStyle": "solid",
  "borderBottomWidth": "1px",
  "borderLeftColor": "black",
  "borderLeftStyle": "dotted",
  "borderLeftWidth": "2px",
  "borderRightColor": "grey",
  "borderRightStyle": "solid",
  "borderRightWidth": "1px",
  "borderTopColor": "grey",
  "borderTopStyle": "solid",
  "borderTopWidth": "1px",
}
`;

exports[`Expanding (with merge) style objects should expand and merge multiple border properties in nested objects 1`] = `
Object {
  "borderBottomColor": "grey",
  "borderBottomStyle": "solid",
  "borderBottomWidth": "7px",
  "borderLeftColor": "black",
  "borderLeftStyle": "dotted",
  "borderLeftWidth": "10px",
  "borderRightColor": "grey",
  "borderRightStyle": "solid",
  "borderRightWidth": "6px",
  "borderTopColor": "grey",
  "borderTopStyle": "solid",
  "borderTopWidth": "5px",
}
`;

exports[`Expanding (with merge) style objects should expand and merge properties in nested objects 1`] = `
Object {
  ":hover": Object {
    "borderBottomWidth": "10px",
    "borderLeftWidth": "4px",
    "borderRightWidth": "2px",
    "borderTopWidth": "5px",
  },
  "paddingBottom": "15px",
  "paddingLeft": "10px",
  "paddingRight": "10px",
  "paddingTop": "20px",
}
`;

exports[`Expanding (with merge) style objects should expand and merge properties with non-undefined values 1`] = `
Object {
  "paddingBottom": 0,
  "paddingLeft": null,
  "paddingRight": undefined,
  "paddingTop": "20px",
}
`;
